File: cibcc/Go/step15/sect2/cibcc.go
package main

import (
	"os"
	"fmt"
)

func main() {
	if len(os.Args) != 2 {
		fmt.Println("引数の個数が正しくありません")
	}

	token = tokenize([]rune(os.Args[1]))
	prog := program()

	for fn := prog.fns; fn != nil; fn = fn.next {
		offset := 0
		for vl := fn.locals; vl != nil; vl = vl.next {
			offset += 8
			vl.va.offset = offset
		}
		fn.stack_size = offset
	}

	codegen(prog)
}

File: cibcc/c++/step27/sect2/cibcc.cpp
#include "cibcc.h"

int align_to(int n, int align);

int main(int argc, char* argv[]) {
	if (argc != 2) {
		std::cout << "引数の個数が正しくありません\n";
	return 1;
	}

	token = tokenize(argv[1]);
	Program *prog = program();
	add_type(prog);

	for (Function *fn = prog->fns; fn; fn = fn->next) {

		int offset = 0;
		for (VarList *vl = fn->locals; vl; vl = vl->next) {
			Var *var = vl->var;
			offset += size_of(var->ty);
			var->offset = offset;
		}
		fn->stack_size = align_to(offset, 8);
	}
	codegen(prog);

	return 0;
}

int align_to(int n, int align) {
	return (n + align - 1) & ~(align - 1);
}
